torch/distributed/device_mesh.py

```diff
@@ -1026,16 +1095,58 @@ def _flatten(
                     "Cannot flatten a DeviceMesh without mesh_dim_names!"
                 )
 
-            if backend_override is not None:
+            if not self._init_backend:
+                raise NotImplementedError(
+                    "flatten a device mesh without backend initialized is not supported!"
+                )
+
+            if not mesh_dim_name:
+                mesh_dim_name = "_".join(not_none(self.mesh_dim_names))
+
+            global_override = _mesh_resources.mesh_dim_group_options.get(
+                0, (None, None)
+            )
+            if backend_override is None:
+                backend_override_tuple = global_override
+            else:
                 (backend_override_tuple,) = _normalize_backend_override(
                     {0: backend_override}, 1
                 )
-            else:
-                backend_override_tuple = (None, None)
+                if backend_override_tuple == (None, None):
+                    backend_override_tuple = global_override
+                elif global_override != (None, None):
+                    raise RuntimeError(
+                        "Dimension 0 present both in the backend_override argument "
+                        "and via _mesh_resources._set_mesh_dim_group_options"
+                    )
+
+            res_sizes, res_strides = zip(
+                *sorted(
+                    [ss for l in self._layouts for ss in l.sizes_and_strides],
+                    key=lambda x: x[1],
+                    reverse=True,
+                )
+            )
+            flattened_layout = _Layout(tuple(res_sizes), tuple(res_strides)).coalesce()
```

Comment:
I think this is the sole direct use of the three support layout functions (coalesce/composition/complement); the others are only done indirectly from local/global ranks